File: bindings/javascript/src/zerotrust.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Zero Trust protocol types for proven-servers.

/** PolicyType matching the Idris2 ABI tags. */
export const PolicyType = Object.freeze({
  ALWAYS_VERIFY: 0,
  NEVER_TRUST: 1,
  LEAST_PRIVILEGE: 2,
  MICRO_SEGMENTATION: 3,
});

/** IdentityConfidence matching the Idris2 ABI tags. */
export const IdentityConfidence = Object.freeze({
  UNVERIFIED: 0,
  BASIC_AUTH: 1,
  MFA_VERIFIED: 2,
  STRONG_AUTH: 3,
  CONTINUOUS_AUTH: 4,
});

/** DeviceTrustScore matching the Idris2 ABI tags. */
export const DeviceTrustScore = Object.freeze({
  DEVICE_UNKNOWN: 0,
  DEVICE_PARTIAL: 1,
  DEVICE_COMPLIANT: 2,
  DEVICE_MANAGED: 3,
  DEVICE_HARDENED: 4,
});

/** AccessDecision matching the Idris2 ABI tags. */
export const AccessDecision = Object.freeze({
  ALLOW: 0,
  DENY: 1,
  CHALLENGE: 2,
  STEP_UP: 3,
});

/** ContextSignalKind matching the Idris2 ABI tags. */
export const ContextSignalKind = Object.freeze({
  LOCATION: 0,
  TIME: 1,
  DEVICE: 2,
  BEHAVIOR: 3,
  NETWORK: 4,
});

/** AuthFactor matching the Idris2 ABI tags. */
export const AuthFactor = Object.freeze({
  CERTIFICATE: 0,
  TOKEN: 1,
  BIOMETRIC: 2,
  FIDO2: 3,
  TOTP: 4,
  PUSH: 5,
});

File: bindings/dart/lib/src/zerotrust.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Zero Trust protocol types for proven-servers.

/// PolicyType matching the Idris2 ABI tags.
enum PolicyType {
  alwaysVerify(0),
  neverTrust(1),
  leastPrivilege(2),
  microSegmentation(3);

  const PolicyType(this.tag);
  final int tag;

  static PolicyType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// IdentityConfidence matching the Idris2 ABI tags.
enum IdentityConfidence {
  unverified(0),
  basicAuth(1),
  mfaVerified(2),
  strongAuth(3),
  continuousAuth(4);

  const IdentityConfidence(this.tag);
  final int tag;

  static IdentityConfidence? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// DeviceTrustScore matching the Idris2 ABI tags.
enum DeviceTrustScore {
  deviceUnknown(0),
  devicePartial(1),
  deviceCompliant(2),
  deviceManaged(3),
  deviceHardened(4);

  const DeviceTrustScore(this.tag);
  final int tag;

  static DeviceTrustScore? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// AccessDecision matching the Idris2 ABI tags.
enum AccessDecision {
  allow(0),
  deny(1),
  challenge(2),
  stepUp(3);

  const AccessDecision(this.tag);
  final int tag;

  static AccessDecision? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ContextSignalKind matching the Idris2 ABI tags.
enum ContextSignalKind {
  location(0),
  time(1),
  device(2),
  behavior(3),
  network(4);

  const ContextSignalKind(this.tag);
  final int tag;

  static ContextSignalKind? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// AuthFactor matching the Idris2 ABI tags.
enum AuthFactor {
  certificate(0),
  token(1),
  biometric(2),
  fido2(3),
  totp(4),
  push(5);

  const AuthFactor(this.tag);
  final int tag;

  static AuthFactor? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
